UM-USElab/gradecraft-development
# Last login bug

### Status
**READY**

### Description
This PR adds two methods that add / update the users last_login_at for themselves (user) and the course they have just logged into. 

This fix updates the user's `last_login_at` attribute, as well as the `user.course_membership.last_login_at` attribute for the current_course they are in.

Five places to test this application are:
* Logging in through GradeCraft with user name and password
* Logging in with Google - Omniauth
* Logging in with Shiboleth - SAML (Testable on staging by visiting http://staging.gradecraft.com/saml/init )
* Logging in through LTI connection 
* Changing a course updates the last_login_at for the user's specific course 


### Todos
- [x] Fix rspec failing test --> course_router_spec.rb line 14 (`Error report -->  Failure/Error: user.update_course_login_at(course.id)
       #<Double :user> received unexpected message :update_course_login_at with (456)`) 

- [x] Add to staging and test logins
- [x] Trace calls to make sure duplicate calls are not being made from login events 
- [x] Make sure we're using update_course_login_at in the proper situations 
- [x] Check in the application where we report `last_activity_at` but call it "Last Login" 

### Migrations
NO

### Steps to Test or Reproduce
Try various ways of logging into the application (5 areas to test noted above) 

then check in the db or on the pages that display "last_login" to see that the users time has been updated. 

### Impacted Areas in Application
users last_login_at

users (current course) last_login_at

======================
Closes #4202
